test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -283,4 +286,140 @@ public void multipleSubscriberCBRTest() throws Exception {
+        //Changing the readergroup from subscriber to non-subscriber
+        readerGroup.resetReaderGroup(nonSubscriberReaderGroupConfig);
+        ReaderGroupConfig.StreamDataRetention readerGroupRetentionType = controller.getReaderGroupConfig(SCOPE_1, READER_GROUP_3).join().getRetentionType();
+        assertEquals(ReaderGroupConfig.StreamDataRetention.NONE, readerGroupRetentionType);
+        assertEquals(0, controller.listSubscribers(SCOPE_1, STREAM_1).join().size());
+
+        // Fill 5 more events to the stream.
+        for (int i = 1; i <= 5; i++) {
+            log.info("Writing event to {}/{}", SCOPE_1, STREAM_1);
+            writer.writeEvent(SIZE_30_EVENT).join();
+        }
+
+        // Retention set has two stream cut at 0/210...0/360
+        // READER_GROUP_3 is not a subscriber as its retention type is updated to NONE
+        // The timeout is set to 3 minutes a little longer than the retention period which is set to 2 minutes
+        // in order to confirm that the retention has taken place.
+        // Check to make sure truncation happened at streamcut in the retention set
+        AssertExtensions.assertEventuallyEquals("Truncation did not take place at offset 210.", true, () -> controller.getSegmentsAtTime(
+                        new StreamImpl(SCOPE_1, STREAM_1), 0L).join().values().stream().anyMatch(off -> off == 210),
+                5000, 3 * 60 * 1000L);
+
+        //Changing the readergroup from non-subscriber to subscriber again
+        readerGroup.resetReaderGroup(readerGroupConfig);
+        readerGroupRetentionType = controller.getReaderGroupConfig(SCOPE_1, READER_GROUP_3).join().getRetentionType();
+        assertEquals(ReaderGroupConfig.StreamDataRetention.MANUAL_RELEASE_AT_USER_STREAMCUT, readerGroupRetentionType);
+        assertEquals(1, controller.listSubscribers(SCOPE_1, STREAM_1).join().size());
+
+        // Recreates the reader
+        reader = clientFactory.createReader(READER_GROUP_3 + "-" + 1, READER_GROUP_3, new JavaSerializer<>(),
+                ReaderConfig.builder().build());
+
+        // fill stream with 3 events
+        for (int i = 1; i <= 3; i++) {
+            log.info("Writing event to {}/{}", SCOPE_1, STREAM_1);
+            writer.writeEvent(SIZE_30_EVENT).join();
+        }
+
+        //Read 3 events with reader from the segment offset
+        for (int i = 1; i <= 3; i++) {
+            read = reader.readNextEvent(READ_TIMEOUT);
+            assertEquals("data of size 30", read.getEvent());
+        }
+
+        log.info("{} generating 2nd stream-cuts for {}/{}", READER_GROUP_3, SCOPE_1, STREAM_1);
+        futureCuts = readerGroup.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update. This will allow for the silent
+        // checkpoint event generated as part of generateStreamCuts to be picked and processed.
+        Futures.delayedFuture(Duration.ofSeconds(5), executor).join();
```

Comment:
Again, wouldn't joining `futureCuts` be enough?

Reply:
Here, readerGroup.generateStreamCuts(streamCutExecutor); generates a silent checkpoint internally which requires a mandatory followup reader.readNextEvent(READ_TIMEOUT) call for the checkpoint to complete and generate a stream cut. If we use futureCuts.join() instead, it waits for the CompletableFuture to complete and keeps on waiting for the checkpoint to complete which does not happen as we could not call readNextEvent. So the delay here is required.